// @api/datastore/index.js
const { Sequelize } = require('sequelize')
let UserModel
let sequelize

class DataStore {
  get provider () {
    return sequelize
  }

  /**
   * Initializes the connection to the database.
   * @param  {String}  connectionString It's a URI...
   * 
   * @return {Promise} 
   */
  async init (connectionString) {
    let myURL

    try {
      myURL = new URL(connectionString)
    } catch {
      throw new Error('the init method requires a connection string parameter that is in the form of a URI')
    }

    sequelize = new Sequelize(connectionString)
  
    UserModel = require('./models/Users.js')(sequelize)
  
    return sequelize.sync()
  }

  /**
   * Creates a new user.
   * @param {Object} newUserInfo 
   * 
   * @returns {Promise} that resolves to a {Number}
   */
  async createUser (newUserInfo) {
    if (sequelize == null) {
      throw new Error('Please call init first')
    }
  
    const user = await UserModel.create({
       firstName: newUserInfo.name.first,
       lastName: newUserInfo.name.last,
       emailAddress: newUserInfo.username,
       password: newUserInfo.password

     })

     return user.id
  }

  async reset () {
    try {
      await sequelize.close()
      sequelize = null
    } catch {
      //safe to ignore 
    }
  }

}

let sharedInstance = null

module.exports = () => {
  if (sharedInstance === null) {
    sharedInstance = new DataStore()
  }

  return sharedInstance
}
